Wrappers/Python/cil/optimisation/utilities/sampler.py

```diff
@@ -0,0 +1,679 @@
+    For example, to be 99% certain that you have seen all indices, for `n=20` you should take at least 152 samples, `n=50` at least 426 samples. To be more likely than not, for `n=20` you should take 78 samples and `n=50` you should take 228 samples. 
+    In general, we note that for a large number of samples (e.g. `>20*num_indices`), the density of the outputted samples looks more and more uniform. For a small number of samples (e.g. `<5*num_indices`) the user may wish to consider
+    another sampling method e.g. random without replacement, which, when calling `num_indices` samples is guaranteed to draw each index exactly once.  
+
+    """
+
+    @staticmethod
+    def sequential(num_indices):
+        """
+        Function that outputs a sampler that outputs sequentially. 
+
+        Parameters
+        ----------
+        num_indices: int
+            The sampler will select from a list of indices {0, 1, …, S-1} with S=num_indices. 
+
+        Example
+        -------
+
+        >>> sampler=Sampler.sequential(10)
+        >>> print(sampler.get_samples(5))
+        >>> for _ in range(11):
+                print(sampler.next())
+
+        [0 1 2 3 4]
+        0
+        1
+        2
+        3
+        4
+        5
+        6
+        7
+        8
+        9
+        0
+        """
+        order = list(range(num_indices))
+        sampler = SamplerFromOrder(num_indices, sampling_type='sequential', order=order, prob_weights=[
+                                   1/num_indices]*num_indices)
+        return sampler
+
+    @staticmethod
+    def custom_order(num_indices, custom_list, prob_weights=None):
+        """
+        Function that outputs a sampler that outputs from a list, one entry at a time before cycling back to the beginning. 
+
+        Parameters
+        ----------
+        num_indices: `int`
+            The sampler will select indices for `{1,....,n}` according to the order in `custom_list` where `n` is `num_indices`. 
+        custom_list: `list` of `int`
+            The list that will be sampled from in order. 
+
+        prob_weights: list of floats of length num_indices that sum to 1. Default is None and the prob_weights are calculated automatically. 
+            Consider that the sampler is called a large number of times this argument holds the expected number of times each index would be called,  normalised to 1. 
+
+        Example
+        --------
+
+        >>> sampler=Sampler.custom_order(12,[1,4,6,7,8,9,11])
+        >>> print(sampler.get_samples(11))
+        >>> for _ in range(9):
+        >>>     print(sampler.next())
+        >>> print(sampler.get_samples(5)) 
+
+        [ 1  4  6  7  8  9 11  1  4  6  7]
+        1
+        4
+        6
+        7
+        8
+        9
+        11
+        1
+        4
+        [1 4 6 7 8]
+
+        """
+        if prob_weights is None:
+            temp_list = []
+            for i in range(num_indices):
+                temp_list.append(custom_list.count(i))
```

Comment:
The idea is that if I have [1,1,6,3,5] and the num_indices=8 it will output [0,2,0,1,0,1,1,0] i.e. 0 zeros, 2 ones, 0 twos, 1 three, 0 fours, 1 five, 1 six and 0 sevens? Does that make sense? Where did [2,2,1,1,1] come from?

Reply:
Ah of course :D I was thinking you were using the custom list value as i.